clang/lib/CIR/CodeGen/CIRGenBuilder.h

```diff
@@ -33,6 +33,14 @@ class CIRGenBuilderTy : public cir::CIRBaseBuilderTy {
       llvm_unreachable("NYI: PPC double-double format for long double");
     llvm_unreachable("Unsupported format for long double");
   }
+
+  bool isSized(mlir::Type ty) {
+    if (mlir::isa<cir::PointerType, cir::ArrayType, cir::BoolType,
+                  cir::IntType>(ty))
+      return true;
+    assert(0 && "Unimplemented size for type");
```

Comment:
looks like we've been doing `llvm_unreachable` for not yet implemented.  That said, its pretty weird that asking this question (`isSized`) is a not-implemented?  What sort of stuff is missing here that we cannot just answer this question?

Reply:
I think it's that the types are not implemented, so we can't return an answer for a type that doesn't exist yet.

Reply:
I am thinking to make `isSized` simple 

`return mlir::isa<cir::PointerType, cir::ArrayType, cir::BoolType, cir::IntType>(ty)`

And when I call it I can assert or errorNYI

Reply:
I think it's better to have the "NYI" indication in the `isSize` definition since that's where the code will need to be updated to add support for other types.

Reply:
I think if we put assert on isSize we can't recover from calling isSize, for example currently I recovering by putting sint32 and 0 size 🤔, what do you think?

Reply:
I like the idea of having `isSized` do an NYI error, then returning a 'sane' value (like false) instead of the assert.

That way, we don't really have to do any real checking in its use, just call 'isSized' or `assert(isSized())` in the use.

Reply:
Hmm... that is unfortunate.  It wouldn't be novel to have it take CGM as an argument, but I don't know what appetite there is for that.

Reply:
`It wouldn't be novel to have it take CGM as an argument` exacily, I think in future we should move `errorNTI` to be a helper function that takes DiagnosticEngine as argument and from any CIEGen we can call it, even in `CIEGenBuilder` we can get Diag from context

Reply:
Still is an assert

Reply:
I think an assert is appropriate here. Perhaps the wording could be changed. This should be returning an answer for every type that we expect to get here. We don't expect non-CIR mlir types, like mlir::Integer.
```suggestion
    assert(0 && "Unexpected MLIR type");
```

Reply:
I think it isn't quite that.... there is an NYI state here as well.  If we dont' expect non-CIR MLIR types, an unreachable would be appropriate here instead.  IF all CIR-MLIR types are sized, asking this question seems silly.

Reply:
Hmm. I just looked again at the incubator implementation. I thought vector types returned false here, but they return `isSized(elementType)`, so nothing implemented there returns false without going through the assert. I compared the incubator code to the classic codegen, and the classic codegen is calling llvm::Type::isSized. Looking into that, it seems to be things like functions, labels, and target extension types that return false. So, you're right that there is something not yet implemented here, but the thing that isn't implemented is the existence of types that need special handling, so we can't even check for this being the thing that isn't implemented.

Maybe the thing to do here is get rid of `CIRGenBuilder::isSized()` completely and add `assert(!cir::MissingFeature::NonSizedTypes());` at the call site. This seems to only be used in two places in the incubator and both will have the same issue.

Reply:
I could live with that.